components/sync15/src/state.rs

```diff
@@ -69,57 +69,81 @@ impl Default for PersistedGlobalState {
     }
 }
 
-#[derive(Debug, Default)]
+#[derive(Debug, Default, Clone, PartialEq)]
 pub(crate) struct EngineChangesNeeded {
     pub local_resets: HashSet<String>,
     pub remote_wipes: HashSet<String>,
 }
 
-impl PersistedGlobalState {
-    pub(crate) fn update_engine_state(
-        &mut self,
-        remote_declined: Option<&[String]>,
-        local_changes: Option<&HashMap<String, bool>>,
-    ) -> (EngineChangesNeeded, bool) {
-        let initial_declined = self.get_declined().to_vec();
-        let lc = local_changes.cloned().unwrap_or_default();
-
-        let mut next_declined: HashSet<String> = remote_declined
-            .unwrap_or(&initial_declined)
-            .iter()
-            .cloned()
-            .collect();
-
-        let mut remote_wipes: HashSet<String> = Default::default();
-
-        for (name, enabled) in lc.iter() {
-            if !*enabled {
-                remote_wipes.insert(name.to_owned());
-            } else {
-                next_declined.remove(name);
-            }
-        }
-
-        for already_declined in remote_declined.unwrap_or_default().iter() {
-            remote_wipes.remove(already_declined);
-        }
+#[derive(Debug, Default, Clone, PartialEq)]
+struct RemoteEngineState {
+    info_collections: HashSet<String>,
+    declined: HashSet<String>,
+}
 
-        let mut newly_declined = next_declined.clone();
-        for e in &initial_declined {
-            newly_declined.remove(e);
-        }
+#[derive(Debug, Default, Clone, PartialEq)]
+struct EngineStateInput {
+    local_declined: HashSet<String>,
+    remote: Option<RemoteEngineState>,
+    user_changes: HashMap<String, bool>,
+}
 
-        self.set_declined(next_declined.into_iter().collect());
+#[derive(Debug, Default, Clone, PartialEq)]
+struct EngineStateOutput {
+    // The new declined.
+    declined: HashSet<String>,
+    // Which engines need resets or wipes.
+    changes_needed: EngineChangesNeeded,
+}
 
-        (
-            EngineChangesNeeded {
-                remote_wipes,
-                local_resets: newly_declined,
-            },
-            !lc.is_empty(),
-        )
+fn compute_engine_states(input: EngineStateInput) -> EngineStateOutput {
+    use crate::util::*;
+    log::debug!("compute_engine_states: input {:?}", input);
+    let (must_enable, must_disable) = partition_by_value(&input.user_changes);
+    let have_remote = input.remote.is_some();
+    let RemoteEngineState {
+        info_collections,
+        declined: remote_declined,
+    } = input.remote.clone().unwrap_or_default();
+
+    let both_declined_and_remote = set_intersection(&info_collections, &remote_declined);
```

Comment:
I didn't know those existed. I implemented the wrappers in terms of them, since they still make things clearer by reducing the `.cloned().collect()` noise.